activerecord/lib/active_record/associations/has_many_through_association.rb

```diff
@@ -38,10 +38,12 @@ def concat_records(records)
       def insert_record(record, validate = true, raise = false)
         ensure_not_nested
 
-        if raise
-          record.save!(:validate => validate)
-        else
-          return unless record.save(:validate => validate)
+        if !record.readonly?
```

Comment:
Another option is
```rb
if record.new_record? || record.changed?
```

That way, if you change a readonly record, we’ll still try to save it and properly blow up. And if the association has `autosave: true` we’ll still save all changes.

We’d also save all changes even if autosave is off – but we do that already. Which is, to quote d849f42b4ecf687ed5350f5a2402fb795aa33aac, “consistently quirky” :-)